### PR TITLE
Feature/2632 logging api

### PR DESCRIPTION
**Pull request by tmandys**

### CODE REVIEW

**Comment by davidnich** [comment]
@tmandys as reported via email, LoggerProgram.qtest is freezing when calling `       m_processing.waitForZero();`, please fix, as it's also causing problems in jenkins (jenkins has to be manually stopped for each bbuild of this branch & blocks all other builds in the meantime)

File: qlib/Logger.qm [hunk 1]
```diff
@@ -0,0 +1,2717 @@
...
+    ....
+    l.info("hello world");
+    l.error("the world is not perfect");
+    @endcode
+
+    @par Multiple Threads
+
+    The user code will log from multiple threads; the events are pushed to a
+    @ref Logger::LoggerAppenderQueue "LoggerAppenderQueue" object. The processing is done in a dedicated thread
+    when the events are passed to appenders. The the user code logging command is unblocking as it terminates
```
Comment (davidnich) [inline qlib/Logger.qm:81]: "The the" and "unblocking" -> "non-blocking"

File: qlib/Logger.qm [hunk 2]
```diff
@@ -0,0 +1,2717 @@
...
+
+    while (True) {
+        laf.getQueue().process();
+    }
+    @endcode
+
+    @par Application Server Running Logging From a Few Sandboxed Program Containers
+
+    In this example, the appserver provides a logger API for a few sandboxed programs. The appserver is
+    responsible for the <b>Log4q</b> configuration; i.e. it prepares loggers, appenders, filters. etc.
```
Comment (davidnich) [inline qlib/Logger.qm:114]: filters, etc

File: qlib/Logger.qm [hunk 3]
```diff
@@ -0,0 +1,2717 @@
...
+    according to configuration and provides the @ref Logger instance to the @ref Qore::Program "Program"
+    container running the sandboxed code.\n
+    The sandboxed code will log to this instance; the logging events are processed by the appserver in a
+    dedidated thread which gets the event from a queue and passes it to appenders.  Multiple loggers may be
```
Comment (davidnich) [inline qlib/Logger.qm:118]: dedicated

File: qlib/Logger.qm [hunk 4]
```diff
@@ -0,0 +1,2717 @@
...
+            @param level The event level
+            @param message The payload as format for @ref sprintf()
+            @param args The arguments for formatting
+            @param locationInfo code location related to logging event
+            @param threadId thread related to the event, Default: @ref gettid()
+            @param timeStamp logging event timestamp. Default value is @ref now_us
+            @param throwable The throwable @ref Qore::ExceptionInfo "ExceptionInfo" associated with logging event
+        */
+        constructor(Logger logger, LoggerLevel level, string message, *list args, *hash<CallStackInfo> locationInfo = NOTHING,
+                    int threadId = gettid(), date timeStamp = now_us(), *hash<ExceptionInfo> throwable = NOTHING) {
```
Comment (davidnich) [inline qlib/Logger.qm:442]: I don't know if we have a standard for this, but I dislike using `= NOTHING` in Qore code - I prefer to use leave it out, as an explicit assignment to no value when this is the default in any case seems to be redundant & does not provide any new information or clarity to the code.  Let's discuss in a dev review meeting

File: qlib/Logger.qm [hunk 5]
```diff
@@ -0,0 +1,2717 @@
...
+        }
+
+        #! Returns the current number of events waiting in queue
+        public int size() {
+            return m_queue.size();
+        }
+
+        #! Returns the next event from queue or @ref nothing if there is no event available within the timeout period
+        /**
+            @param ms a timeout value to wait for data to become available on the queue;
```
Comment (davidnich) [inline qlib/Logger.qm:656]: I realize that the semantics for the timeout come from `Queue::get()`, but these are really unfortunate, as they are counterintuitive.  Maybe we should consider implementing 0 = immediate timeout, < 0 = infinite timeout here (and other places that call `getEvent()` - what do you think?

File: qlib/Logger.qm [hunk 6]
```diff
@@ -0,0 +1,2717 @@
...
+                } catch (hash<ExceptionInfo> ex) {
+                    switch (ex.err) {
+                        case "QUEUE-TIMEOUT":
+                            break;
+                        default:
+                            rethrow;
+                    }
+                }
+            }
+            return NOTHING;
```
Comment (davidnich) [inline qlib/Logger.qm:679]: same comment here as with assigning to no value; returning an explicit "no value" seems redundant and also has a performance cost - particularly in system APIs we should be optimized for maximum performance

File: qlib/Logger.qm [hunk 7]
```diff
@@ -0,0 +1,2717 @@
...
+
+        #! worker thread code
+        /**
+            @param id unique identifier generated by internal @ref Qore::Thread::Sequence "Sequence"
+            @param appender @ref LoggerAppender instance
+            @param events list of events
+        */
+        private:internal worker(string id, LoggerAppender appender, list events) {
+            try {
+                foreach hash e in (events) {
```
Comment (davidnich) [inline qlib/Logger.qm:719]: this would be faster as
```
map appender.processEventImpl($1.type, $1.params), events;
```

File: qlib/Logger.qm [hunk 8]
```diff
@@ -0,0 +1,2717 @@
...
+
+            # remove finished
+            while (m_finished.size() > 0) {
+                string id = m_finished.get(-1);
+                remove m_processing{id};
+            }
+            # now try push pending
+            foreach string id in (keys m_pending) {
+                # remove finished
+                #while (m_finished.size() > 0) {
```
Comment (davidnich) [inline qlib/Logger.qm:806]: please remove this code if it's not necessary

File: qlib/Logger.qm [hunk 9]
```diff
@@ -0,0 +1,2717 @@
...
+                if (m_maxThreads >= 0 && m_processing.size() >= m_maxThreads) {
+                    return;
+                }
+                if (exists m_processing{id}) {
+                    # we cannot trigger new thread if is being processed
+                    continue;
+                }
+                hash e = remove m_pending{id};
+                m_processing{id} = e;
+                # BUG: when passing "id" then no copy is passed to submit pool queue and worker may get wrong value
```
Comment (davidnich) [inline qlib/Logger.qm:819]: I don't understand the bug here, please explain - also what does "real constant enough value" mean?

File: qlib/Logger.qm [hunk 10]
```diff
@@ -0,0 +1,2717 @@
...
+        #! Creates the object
+        /**
+            @param name the appender name
+        */
+        constructor(string name = "") {
+            m_name = name;
+        }
+
+        destructor() {
+            #close();  # TODO: cannot call as the target object may be already destroyed
```
Comment (davidnich) [inline qlib/Logger.qm:876]: let's discuss this - it looks like this issue needs attention before merging

File: qlib/Logger.qm [hunk 11]
```diff
@@ -0,0 +1,2717 @@
...
+                if (i >= s.size()) {
+                    push pp, s;
+                    s = "";
+                } else {
+                    if (i > 0) {
+                        push pp, substr(s, 0, i);
+                    }
+                    splice s, 0, i+1;
+                    # now the string is token
+%ifdef _DUMMY_TO_PAIR_BRACES_IN_EDITOR
```
Comment (davidnich) [inline qlib/Logger.qm:1151]: what is this for?  it sounds like you need to improve your editor's qore support or switch to another editor (VSCode for example?)

File: qlib/Logger.qm [hunk 12]
```diff
@@ -0,0 +1,2717 @@
...
+            m_pattern = pattern;
+            m_parsedPattern = pp;
+        }
+
+        #! Returns the current pattern
+        public string getPattern() {
+            return m_pattern;
+        }
+
+        #! Returns the parsed pattern as a list
```
Comment (davidnich) [inline qlib/Logger.qm:1196]: you could solve this without this methog by making `m_parsedPattern` `private` and subclassing this class in the test for example

**Review by davidnich — CHANGES_REQUESTED** [review]

File: qlib/Logger.qm (hunk 11 above)
Comment (tmandys) [reply qlib/Logger.qm:1151]: At least vim and kate are confused with odd right curly bracket in regex and fail block pairing in file. I think pairing is editor buil- in function

File: qlib/Logger.qm (hunk 5 above)
Comment (tmandys) [reply qlib/Logger.qm:656]: Question if is not more counterintuitive two timeout logics for two methods.

Comment (davidnich) [reply qlib/Logger.qm:656]: `Queue::get()` is the outlier - everywhere else 0 = no timeout and <0 = infinite timeout - such as for example `Mutex::lock()`

File: qlib/Logger.qm [hunk 13]
```diff
@@ -0,0 +1,2717 @@
...
+                        res = sprintf(option ?? "%d", data.count);
+                        break;
+                    case "P":
+                        res = data.filename;
+                        break;
+                    case "p": # path
+                        res = dirname(data.filename);
+                        if (res != DirSep) {
+                            res += DirSep;
+
```
Comment (davidnich) [inline qlib/Logger.qm]: docs need to be updated with an example using variable args - also other similar methods

**Review by davidnich — CHANGES_REQUESTED** [review]

**Pull request closed without merging**